qubes/app.py

```diff
@@ -1404,6 +1404,24 @@ def _get_pool(**kwargs):
             raise qubes.exc.QubesException('No driver %s for pool %s' %
                                            (driver, name))
 
+    async def stop_storage(self, loop):
+        """
+        Stop the storage of all domains that are not running.
+        """
+        async def stop(i):
+            async with i.startup_lock:
+                if not i.is_running():
+                    await i.storage.stop()
+        future = (qubes.utils.void_coros_maybe(stop(i)
+            for i in self.domains if i.klass != 'AdminVM'),)
+        finished = ()
+        while future:
+            qubes.utils.systemd_extend_timeout()
+            finished, future = await asyncio.wait(
+                future, loop=loop, timeout=30000000)
+        for i in finished:
```

Comment:
This awaits only tasks from the last iteration (you override `finished` variable).

Reply:
`asyncio.wait` splits the futures passed to it between its two return values.  Here, `future` starts as a 1-tuple, so when `finished` becomes nonempty, `future` will become empty and the loop will exit.  Therefore, I believe this code is currently correct as is, but it should certainly be made clearer, and removing the use of `void_coros_maybe` will require it to be fixed.

Reply:
Ah, indeed the `void_coros_maybe` was called on a generator, not on each `stop(i)` call separately.